crates/pallet-executor/src/lib.rs

```diff
@@ -586,6 +596,58 @@ impl<T: Config> Pallet<T> {
         Ok(())
     }
 
+    fn validate_bundle_election(
+        executor_id: &ExecutorId,
+        proof_of_election: &ProofOfElection,
+    ) -> Result<(), BundleError> {
+        let ProofOfElection {
+            domain_id,
+            vrf_output,
+            vrf_proof,
+            vrf_public_key,
+            slot_randomness,
+            state_root,
+            storage_proof,
+        } = proof_of_election;
+
+        verify_vrf_proof(vrf_public_key, vrf_output, vrf_proof, slot_randomness)
+            .map_err(|_| BundleError::BadVrfProof)?;
+
+        // TODO: verify `state_root` is valid.
+
+        let BundleElectionParams {
+            authorities,
+            total_stake_weight,
+            slot_probability,
+        } = read_bundle_election_params(storage_proof.clone(), state_root)
+            .map_err(|_| BundleError::BadStorageProof)?;
+
+        let stake_weight = authorities
+            .iter()
+            .find_map(|(authority, weight)| {
+                if authority == executor_id {
+                    Some(weight)
+                } else {
+                    None
+                }
+            })
+            .ok_or(BundleError::AuthorityNotFound)?;
+
+        let election_solution = derive_bundle_election_solution(*domain_id, vrf_output);
+
+        let threshold = calculate_bundle_election_threshold(
+            *stake_weight,
+            total_stake_weight,
+            slot_probability,
+        );
+
+        if u128::from(election_solution) > threshold {
```

Comment:
In farmer code what I did is abstract this away into a single function that returns `true` if within threshold, seems to be safer that way.